odl/test/discr/discr_utils_test.py

```diff
@@ -0,0 +1,265 @@
+
+from __future__ import division
+
+import numpy as np
+import pytest
+
+import odl
+from odl.discr.discr_utils import (
+    linear_interpolator, nearest_interpolator, per_axis_interpolator,
+    point_collocation)
+from odl.discr.grid import sparse_meshgrid
+from odl.util.testutils import all_almost_equal, all_equal
+
+
+def test_nearest_interpolation_1d_complex():
+    """Test nearest neighbor interpolation in 1d with complex values."""
+    coord_vecs = [[0.1, 0.3, 0.5, 0.7, 0.9]]
+    f = np.array([0 + 1j, 1 + 2j, 2 + 3j, 3 + 4j, 4 + 5j], dtype="complex128")
+    interpolator = nearest_interpolator(f, coord_vecs)
+
+    # Evaluate at single point
+    val = interpolator(0.35)  # closest to index 1 -> 1 + 2j
+    assert val == 1.0 + 2.0j
+    # Input array, with and without output array
+    pts = np.array([0.39, 0.0, 0.65, 0.95])
+    true_arr = [1 + 2j, 0 + 1j, 3 + 4j, 4 + 5j]
+    assert all_equal(interpolator(pts), true_arr)
+    # Should also work with a (1, N) array
+    pts = pts[None, :]
+    assert all_equal(interpolator(pts), true_arr)
+    out = np.empty(4, dtype='complex128')
+    interpolator(pts, out=out)
+    assert all_equal(out, true_arr)
+    # Input meshgrid, with and without output array
+    # Same as array for 1d
+    mg = sparse_meshgrid([0.39, 0.0, 0.65, 0.95])
+    true_mg = [1 + 2j, 0 + 1j, 3 + 4j, 4 + 5j]
+    assert all_equal(interpolator(mg), true_mg)
+    interpolator(mg, out=out)
+    assert all_equal(out, true_mg)
+
+
+def test_nearest_interpolation_2d():
+    """Test nearest neighbor interpolation in 2d."""
+    coord_vecs = [[0.125, 0.375, 0.625, 0.875], [0.25, 0.75]]
+    f = np.array([0, 1, 2, 3, 4, 5, 6, 7], dtype="float64").reshape((4, 2))
+    interpolator = nearest_interpolator(f, coord_vecs)
+
+    # Evaluate at single point
+    val = interpolator([0.3, 0.6])  # closest to index (1, 1) -> 3
+    assert val == 3.0
+    # Input array, with and without output array
+    pts = np.array([[0.3, 0.6],
+                    [1.0, 1.0]])
+    true_arr = [3, 7]
+    assert all_equal(interpolator(pts.T), true_arr)
+    out = np.empty(2, dtype='float64')
+    interpolator(pts.T, out=out)
+    assert all_equal(out, true_arr)
+    # Input meshgrid, with and without output array
+    mg = sparse_meshgrid([0.3, 1.0], [0.4, 1.0])
+    # Indices: (1, 3) x (0, 1)
+    true_mg = [[2, 3],
+               [6, 7]]
+    assert all_equal(interpolator(mg), true_mg)
+    out = np.empty((2, 2), dtype='float64')
+    interpolator(mg, out=out)
+    assert all_equal(out, true_mg)
+
+
+def test_nearest_interpolation_2d_string():
+    """Test nearest neighbor interpolation in 2d with string values."""
+    coord_vecs = [[0.125, 0.375, 0.625, 0.875], [0.25, 0.75]]
+    f = np.array(list('mystring'), dtype='U1').reshape((4, 2))
+    interpolator = nearest_interpolator(f, coord_vecs)
+
+    # Evaluate at single point
+    val = interpolator([0.3, 0.6])  # closest to index (1, 1) -> 3
+    assert val == u't'
```

Comment:
What is the u for?

Reply:
To make sure it's unicode. Maybe the `==` comparison would also work without, but this seems safer for Python 2.